File: src/core/schema/tools/equals.rb
#=begin

#Deep equality based on object tree structure and primitive values

#=end
require 'enso'

module Equals
class EqualsClass

  def initialize()
    @memo = {}
  end
  
  def self.equals(a, b)
    self.new.equals(a, b)
  end

  def equals(a, b)
    if a == b
      true
    elsif a.nil? || b.nil? || a.schema_class.name != b.schema_class.name
      false
    elsif @memo[[a, b]]
      true
    else
      res = true
      @memo[[a, b]] = true

      a.schema_class.fields.each do |field|
        a_val = a[field.name]
        b_val = b[field.name]
        if field.type.is_a?("Primitive")
          res = false if a_val != b_val
        elsif !field.many
          if !equals(a_val, b_val)
            puts "fail2 #{a_val} #{b_val}"
            res = false
          end
        elsif a_val.is_a?(Factory::List)
          res = false if !equals_list(a_val, b_val)
        elsif a_val.is_a?(Factory::Set)
          res = false if !equals_set(a_val, b_val)
        end
      end
      res
    end
  end

  def equals_list(l1, l2)
    if l1.size!=l2.size
      false
    else
      res = true
      l1.keys.each do |i|
        res = false if !equals(l1[i], l2[i])
      end
      res
    end
  end

  def equals_set(l1, l2) #like equals list but ignores order
    if l1.size!=l2.size
      false
    else
      res = true
      l1.keys.each do |i|
        res = false if !equals(l1[i], l2[i])
      end
      res
    end
  end
end

def self.equals(a,b)
  EqualsClass.equals(a, b)
end

end

File: src/js/core/schema/tools/equals.js
'use strict'

//// Equals ////

var cwd = process.cwd() + '/';
var Enso = require(cwd + "enso.js");

var Equals;

var equals = function(a, b) {
  var self = this;
  return EqualsClass.equals(a, b);
};

class EqualsClass {
  static new(...args) { return new EqualsClass(...args) };

  static equals(a, b) {
    var self = this;
    return self.new().equals(a, b);
  };

  constructor() {
    var self = this;
    self.memo$ = Enso.EMap.new();
  };

  equals(a, b) {
    var self = this, res, a_val, b_val;
    if (a == b) {
      return true;
    } else if ((a == null || b == null) || a.schema_class().name() != b.schema_class().name()) {
      return false;
    } else if (self.memo$.get$([a, b])) {
      return true;
    } else {
      res = true;
      self.memo$ .set$([a, b], true);
      a.schema_class().fields().each(function(field) {
        a_val = a.get$(field.name());
        b_val = b.get$(field.name());
        if (Enso.System.test_type(field.type(), "Primitive")) {
          if (a_val != b_val) {
            return res = false;
          }
        } else if (! field.many()) {
          if (! self.equals(a_val, b_val)) {
            puts(Enso.S("fail2 ", a_val, " ", b_val));
            return res = false;
          }
        } else if (Enso.System.test_type(a_val, Factory.List)) {
          if (! self.equals_list(a_val, b_val)) {
            return res = false;
          }
        } else if (Enso.System.test_type(a_val, Factory.Set)) {
          if (! self.equals_set(a_val, b_val)) {
            return res = false;
          }
        }
      });
      return res;
    }
  };

  equals_list(l1, l2) {
    var self = this, res;
    if (l1.size_M() != l2.size_M()) {
      return false;
    } else {
      res = true;
      l1.keys().each(function(i) {
        if (! self.equals(l1.get$(i), l2.get$(i))) {
          return res = false;
        }
      });
      return res;
    }
  };

  equals_set(l1, l2) {
    var self = this, res;
    if (l1.size_M() != l2.size_M()) {
      return false;
    } else {
      res = true;
      l1.keys().each(function(i) {
        if (! self.equals(l1.get$(i), l2.get$(i))) {
          return res = false;
        }
      });
      return res;
    }
  };
};

Equals = {
  equals: equals,
  EqualsClass: EqualsClass,
};
module.exports = Equals ;
